source/adaptive_load/scoring_function_impl.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include "envoy/registry/registry.h"
+
+#include "nighthawk/adaptive_load/scoring_function.h"
+
+#include "api/adaptive_load/scoring_function_impl.pb.h"
+
+namespace Nighthawk {
+
+/**
+ * ScoringFunction that returns 1.0 when a metric is within thresholds and -1.0 otherwise.
+ * Supports upper or lower threshold or both.
+ */
+class BinaryScoringFunction : public ScoringFunction {
+public:
+  explicit BinaryScoringFunction(
+      const nighthawk::adaptive_load::BinaryScoringFunctionConfig& config);
+  double EvaluateMetric(double value) const override;
+
+private:
+  // Upper threshold for the metric.
+  const double upper_threshold_;
+  // Lower threshold for the metric.
+  const double lower_threshold_;
+};
+
+/**
+ * Factory that creates a BinaryScoringFunction from a BinaryScoringFunctionConfig proto.
+ * Registered as an Envoy plugin.
+ */
+class BinaryScoringFunctionConfigFactory : public ScoringFunctionConfigFactory {
+public:
+  std::string name() const override;
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+  ScoringFunctionPtr createScoringFunction(const Envoy::Protobuf::Message& message) override;
+  absl::Status ValidateConfig(const Envoy::Protobuf::Message& message) const override;
+};
+
+// This factory is activated through LoadScoringFunctionPlugin in plugin_util.h.
+DECLARE_FACTORY(BinaryScoringFunctionConfigFactory);
+
+/**
+ * ScoringFunction that calculates a metric score as k * (threshold - value), where k is a scaling
+ * constant. The score is 0.0 when the value exactly equals the threshold, positive below the
+ * threshold (meaning load should increase), and negative above the threshold. The score is
+ * proportional to the difference from the threshold.
+ */
+class LinearScoringFunction : public ScoringFunction {
+public:
+  explicit LinearScoringFunction(
+      const nighthawk::adaptive_load::LinearScoringFunctionConfig& config);
+  double EvaluateMetric(double value) const override;
```

Comment:
I was thinking of a using a class enum and then if/else in the implementation that uses the magic double for each case.

Reply:
Filed issue #473 to discuss this -- we could potentially refactor this.